AGENT-436: Propose interactive network config for agent-based installer (#1310)
openshift/enhancements: enhancements/agent-installer/interactive-network-config.md

connectivity for this step, configuration will be via a text user interface
(TUI) displayed on the console.

## Motivation

Providing network configuration up-front, in the absence of the actual hosts to
be used, is onerous for users. Names of NIC interfaces are difficult to predict
in advance, and often MAC addresses are required. Even the simplified NMState
format is difficult to write by hand.

Correcting any error in the configuration passed into the installer currently
requires returning to the beginning of the process, generating a new ISO and
booting the hosts anew with the updated image. There is no guarantee of getting
it right, so there could be multiple iterations of this process.

### User Stories

* As a cluster administrator, I want to install a cluster into a network
  environment requiring custom network configuration (e.g. static IPs, NIC
  bonds, VLANs) without needing any information from the physical servers in
  advance.
* As a cluster administrator, I want to fix a cluster installation that has
  stalled due to incorrect network configuration without having to regenerate
  the installation image.

### Goals

> * Allow the user to configure static IPs, NIC bonds, and VLANs.

Reviewer: Should we also include the rendezvous ip?

Author: It's discussed in the alternatives section, but I don't think we should. At least for now.

* Make available all necessary information about the hardware and network
  environemnt to do this configuration.

> * Provide an indication of when sufficient connectivity is or is not available.

Reviewer: I think it could be useful, at this point of the doc, to provide a little bit more details about "sufficient connectivity", ie possibility for an host to reach the rendezvous host, to reach a registry, and so on (even though more info are provided later)

* Enable the user to debug connectivity problems by narrowing down the cause
  (DNS, reachability, &c.).
* Work for both the existing automation workflow and a future fully-interactive
  workflow.

### Non-Goals

* Interactive (re)configuration of any other aspect of the installation.
* Add interactive network configuration to CoreOS in general.
* Allow interactive network configuration during IPI baremetal installs.

## Proposal

### Workflow Description

The user will generate the agent ISO using `openshift-install` as usual. The
`networkConfig` section in the `agent-config` (or the NMStateConfig ZTP
manifests) may _or may not_ contain network configuration.

> The user will then boot the ISO. Any network configuration for each current

Reviewer 2: If the agent-config contains a hostname for the host, it should be applied before bringing up the TUI. Is this correct?

Author: Yes, I think it should. Users would probably be confused if they configured hostnames and then those weren't applied when they started manually editing the config.

host contained in the ISO will be applied at startup.

Prior to the login prompt, a TUI will appear on the hardware console. It will
display the current status of connectivity to the registry, which will be
continuously updated. It will also give the user the option to interactively
reconfigure the network, or accept the current configuration and exit.

> If there is connectivity available to the registry, the TUI will autmatically

Reviewer 3: Nit - typo "autmatically"

> If there is connectivity available to the registry, the TUI will autmatically

Reviewer: ```suggestion
If there is connectivity available to the registry, the TUI will automatically
```

> exit within 20s if there is no user interaction.

Reviewer 3: Is this the same for the "non-host0" hosts? Will there be a way to recover if the issue is that the host0 host is not reachable by other hosts because of incompatible IP config or will that require a 2nd reboot?

Author: I'm proposing not to check for connectivity to node 0, because it limits our options for future development - both for getting this feature into CoreOS and for how we implement the interactive workflow.
In practice I suspect connectivity to the registry from each host is usually going to be sufficient for nodes to find each other. If it isn't then the user will have to log in and fix it or reboot and try again I guess.

Reviewer 3: OK, sounds good

Services that require pulling a container image will not start until the TUI
exits.

Once the TUI exits, the login prompt will be displayed on the console as usual.

### API Extensions

N/A

### Implementation Details/Notes/Constraints

The TUI will be implemented as a golang binary and delivered in a container

> added to the release payload. The actual network configuration will be done by

Reviewer 2: ```suggestion
added to the release payload. The actual network configuration will be done 
```

> through the existing `nmtui` binary, which is already present in CoreOS.

Reviewer 4: So the new tool just shows the connectivity status and lets the user choose to reconfigure the network, then it launches nmtui? If that's right, you could add as a non-goal something like "building a complete TUI for editing network settings to replace nmtui".

To allow it to run before the network is configured, the TUI binary will be
extracted from the release payload by the installer process, and embedded in
the CoreOS ISO (which is also extracted from the release payload). Since even a
modest golang binary is likely too large for the pre-allocated ignition area,
this will be added to the ISO either as a separate volume mount or as an
additional CPIO archive appended to a rewritten ignition volume. In PXE

> installs, it will be in an additional CPIO archive appended to the rootfs.

Reviewer 4: This implies the embedding may be done in different ways under different conditions. Is that necessary, or could we always do it one way?

Do we have any idea how big this binary is likely to be?

Reviewer: From the preliminary work done in the PoC for embedding a binary in the ISO, currently the TUI binary is ~17Mb. The ignition.img (which is a compressed cpio archive) goes up to ~7Mb

Reviewer 4: OK, I was commenting on this part: "this will be added to the ISO either as a separate volume mount or as an additional CPIO archive appended to a rewritten ignition volume"

Maybe that indicates the decision about the approach hasn't been made? When I read it the first time, I thought it meant both approaches might be used in different circumstances.

Author: Oh, I thought you were commenting on the PXE part.
For the ISO, yeah it's still one of the open questions (hence 'either').
For PXE it might be different.

The running network configuration is copied to disk using the `--copy-network`
argument to `coreos-installer`, so no other changes are required to persist the
desired configuration to the network.

Access to the registry could be checked by simply telling podman to pull the
release image. However, if this fails then ideally we would be able to identify
the point where it stops working in the following steps:

* DNS lookup
* ping
* HTTP
* Registry login

> * Pulling the release image

Reviewer 3: In a disconnected environment it could also be an issue with the ICSP configuration. Would it be a step to change to change registries.conf to fix mirror misconfiguration or will that require regenerating the ISO?

Author: I think that would require regeneration, since it's used in multiple places (not just `registries.conf`, but also in the ZTP manifests. Also, we want it to be the same across hosts.

### Risks and Mitigations

Users with access to the hardware console will be able to manipulate the
network configuration without logging in. If an exploit is found they may be

> able to upgrade this to shell access. To mitgate this, the service should not

Reviewer 3: Nit - typo "mitgate"

run as the root user. However, a user with physical access to the hardware or

> access to the BMC can probably do anything they want anyway.

Reviewer 4: Can the service update the networking files if it isn't running with root privileges?

If there is connectivity available to the registry, the TUI will automatically
exit within 20s if there is no user interaction.

Services that require pulling a container image will not start until the TUI
exits.

Once the TUI exits, the login prompt will be displayed on the console as usual.

### API Extensions

N/A

### Implementation Details/Notes/Constraints

The TUI will be implemented as a golang binary and delivered in a container
added to the release payload. The actual network configuration will be done
through the existing `nmtui` binary, which is already present in CoreOS.

To allow it to run before the network is configured, the TUI binary will be
extracted from the release payload by the installer process, and embedded in
the CoreOS ISO (which is also extracted from the release payload). Since even a
modest golang binary is likely too large for the pre-allocated Ignition config
area, the binary will be added to the initramfs along with a script in the
`/usr/lib/dracut/hooks/pre-pivot/` directory to copy it from the initramfs to
the system root. To add these files to the initramfs in the ISO, they will be

> added to a new CPIO archive, and this archive will be appended to the ignition

Reviewer: ```suggestion
added to a new compressed CPIO archive, and this archive will be appended to the initrd
```

volume *after* the empty space at the end. This allows users to continue to use
the `coreos-installer` commands to extract the Ignition config and replace it
with a different one. In PXE installs, the installer will append the new CPIO
archive to the rootfs file, immediately after the CPIO archive containing the
Ignition config.

The running network configuration is copied to disk using the `--copy-network`
argument to `coreos-installer`, so no other changes are required to persist the
desired configuration to the network.

Access to the registry could be checked by simply telling podman to pull the

> release image. However, if this fails then ideally we would be able to identify

Reviewer: To avoid downloading initially the image the Docker registry API could be used as well
